### DISCUSSION TITLE
Steps to build a CSDMS Workbench component

### DISCUSSION THREAD

**Discussion by mdpiper** [csdms/project#19] [General]
Here are the steps to build a component for the CSDMS Workbench from a library containing a model or a tool.

To help explain these steps, consider a library *foo* packaged in a Python module `foo` that exposes a class `Foo`.

1. Wrap the library with a Basic Model Interface (BMI). For the case of the *foo* library, create a class `BmiFoo` and include in the `foo` module or in a separate `bmi_foo` module.
   * Learn more about BMI from its [documentation](https://bmi.readthedocs.io)
   * See examples of wrapping a simple model with a BMI in [C](https://github.com/csdms/bmi-example-c), [C++](https://github.com/csdms/bmi-example-cxx), [Fortran](https://github.com/csdms/bmi-example-fortran), and [Python](https://github.com/csdms/bmi-example-python)
   * See examples of wrapping data sources with [soilgrids](https://soilgrids.readthedocs.io), [nwis](https://nwis.readthedocs.io), and [gridmet](https://github.com/nhm-usgs/gridmet_bmi)
1. [Optional, but recommended] Create a [conda-forge](https://conda-forge.org/) package for the BMI library. For the *foo* library, the conda-forge package could be `foo` or `bmi-foo`.
   * The conda-forge documentation on [contributing packages](https://conda-forge.org/docs/maintainer/adding_pkgs.html) is essential reading
   * Examples of conda-forge packages that wrap a model or data source with a BMI include [hydrotrend](https://github.com/conda-forge/hydrotrend-feedstock), [permamodel](https://github.com/conda-forge/permamodel-feedstock), [gipl](https://github.com/conda-forge/gipl-feedstock), and [gridmet_bmi](https://github.com/conda-forge/gridmet_bmi-feedstock)
1. Run the *babelizer* over the BMI library to create a component that can be imported into *pymt*. Babelizing the `BmiFoo` class would produce the `pymt_foo` module exposing the `Foo` class.
   * Learn more about the *babelizer* from its [documentation](https://babelizer.readthedocs.io), and about *pymt* from its [documentation](https://pymt.readthedocs.io)
   * Examples of components include [pymt_hydrotrend](https://github.com/pymt-lab/pymt_hydrotrend), [pymt_permamodel](https://github.com/pymt-lab/pymt_permamodel), [pymt_gipl](https://github.com/pymt-lab/pymt_gipl), and [pymt_gridmet](https://github.com/pymt-lab/pymt_gridmet)
1. [Optional, but recommended] Create a conda-forge package for the *pymt* component. For example, `pymt_foo`.
   * As mentioned above, the conda-forge documentation on [contributing packages](https://conda-forge.org/docs/maintainer/adding_pkgs.html) is essential reading
   * Examples of conda-forge packages of CSDMS Workbench components include [pymt_hydrotrend](https://github.com/conda-forge/pymt_hydrotrend-feedstock), [pymt_permamodel](https://github.com/conda-forge/pymt_permamodel-feedstock), and [pymt_gipl](https://github.com/conda-forge/pymt_gipl-feedstock)

@mcflugen @gantian127 @gregtucker: Please feel free edit these steps as you see fit, and add any discussion as comments below!

**Comment by gregtucker** [comment]
**Nice!** How about including links for people to learn about the steps, e.g., (Learn about BMI) <= link to bmi.readthedocs.org, and (Learn how to create a conda-forge package) <= link to whatever pages on conda-forge, etc.

**Comment by mdpiper** [comment]
@gregtucker Done! I've added links to docs and examples.
